packages/hardhat-core/src/internal/solidity/parse.ts

```diff
@@ -55,12 +59,28 @@ export class Parser {
     return result;
   }
 
-  private _getFromCache(absolutePath: string): ParsedData | null {
-    const cacheEntry = this._solidityFilesCache?.getEntry(absolutePath);
+  /**
+   * Get parsed data from the internal cache, or from the solidity files cache.
+   *
+   * Returns null if cannot find it in either one.
+   */
+  private _getFromCache(
+    absolutePath: string,
+    contentHash: string
+  ): ParsedData | null {
+    if (this._solidityFilesCache === undefined) {
+      return this._cache.get(absolutePath) ?? null;
```

Comment:
What if we use the content hash instead of the absolute path here? I think it will be more robust. I can imagine using the `absolutePath` bringing issues for things like the `hardhat-watcher` plugin.